### PR TITLE
Ability to define search columns

### PR DESCRIPTION
**Pull request by RFSH**
This PR will implement the changes discussed in [#744 comment](https://github.com/informatics-isi-edu/ermrestjs/issues/744#issuecomment-545063518), which mainly includes:

- Support `sourcekey` in the facet part of URLs.
- Add support for special `search-box` sourcekey, which will be used to manipulate the search functionality. 

These are the summary of changes that I had to make to support this:

- `Table.searchSourceDefinition` that is used in both parser helper function and `Reference`.
- `Reference.searchColumns` that chaise will use to construct the placeholder.
- Add a mechanism to attach the `catalog` object to `Location` object. Since we want to support `sourcekey`, we have to get some information from the table and therefore needed the catalog.
- Refactor the old `_JSONToErmrestFilter ` function:
  - I didn't like the name so I renamed it to `_renderFacet`. It's better but still not ideal. This function will turn a given facet object into its equivalent ermrest URL.
  - Moved all the helper functions outside and scoped it to `_renderFacetHelpers` so it's more apparent what they are used for. I think I should do the same with the rest of the functions that were the same as this function e.g. `Reference.facetColumns` and `PseudoColumn.getAggregatedValue`.
 - Add support for `sourcekey` to ` _renderFacet` function.
 - Test cases:
   - I had to change all the places that were using old search syntax, to the new one.
   - Parser test cases were not using any table, but for the `sourcekey` support test cases I had to add a schema to this spec. I also had to modify the other test cases based on this.

P.S. The only remaining task from the issue is the ability to go beyond the table (defining sources with foreign key paths in the `search-box`) which requires more discussion and should be implemented later.

### CODE REVIEW

File: js/parser.js [hunk 1]
```diff
@@ -153,6 +159,14 @@
         this._catalog = catalogParts[0];
         this._version = catalogParts[1] || null;
 
+        if (catalogObject) {
```
Comment (jrchudy) [inline js/parser.js]: Isn't it the case that the catalog is always used now? Didn't you make changes to each calling site?

File: js/utils/pseudocolumn_helpers.js [hunk 2]
```diff
@@ -215,116 +196,200 @@
...
+     * @param       {string} catalogId the catalog id
+     * @param       {ERMrest.catalog} [catalogObject] the catalog object (could be undefined)
+     * @param       {Object[]} [consNames] the constraint names (could be undefined)
+     * @constructor
+     * @return      {object} An object that will have the following attributes:
+     * - successful: Boolean (true means successful, false means cannot be parsed).
+     * - parsed: String  (if the given string was parsable).
+     * - message: String (if the given string was not parsable)
+     */
+    _renderFacet = function(json, alias, schemaName, tableName, catalogId, catalogObject, consNames) {
```
Comment (jrchudy) [inline js/utils/pseudocolumn_helpers.js]: The diff for this file is a bit goofy. It looks like this is where you changed `_JSONToErmrestFilter` to `_renderFact`. It looks like you also moved functions that were scoped within that function to be outside of it for reuse. I don't see many changes to the actual logic though besides explicitly checking for the `SEARCH_BOX` source key.

File: js/parser.js (hunk 1 above)
Comment (RFSH) [reply js/parser.js]: Internally in ermrestjs, this function is used without passing the `catalog` object. For example, the `resolve` function is using this to figure out which catalog it should use, so it doesn't have access to the `catalogObject` yet and cannot pass it.

File: js/utils/pseudocolumn_helpers.js (hunk 2 above)
Comment (RFSH) [reply js/utils/pseudocolumn_helpers.js]: That's right. The only actual change to this function is supporting `sourcekey` which includes the `SEARCH_BOX` support.

**Pull request merged**